### PR TITLE
feat(keystone): INFRA-265 rate limit observer proof of work

### PR DESCRIPTION
**Pull request by SavelevMatthew**

### CODE REVIEW

File: packages/keystone/rateLimiting/constants.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+const DEFAULT_MUTATION_WEIGHT = 5
+const DEFAULT_QUERY_WEIGHT = 1
+const DEFAULT_MAX_TOTAL_RESULTS = 1000
+const DEFAULT_AUTHED_QUOTA = 5000
+const DEFAULT_NON_AUTHED_QUOTA = 100
+const DEFAULT_QUOTA_WINDOW = '1h'
+const DEFAULT_WHERE_COMPLEXITY_FACTOR = 2
+const DEFAULT_PAGE_LIMIT = 100
```
Comment (sitozzz) [inline packages/keystone/rateLimiting/constants.js:8]: Why all these constants starts with `DEFAULT_` prefix? Is there any option to override it's values?
If so, I would prefer to see the following:
```suggestion
const MUTATION_WEIGHT = 5
const QUERY_WEIGHT = 1
const MAX_TOTAL_RESULTS = 1000
const AUTHED_QUOTA = 5000
const NON_AUTHED_QUOTA = 100
const QUOTA_WINDOW = '1h'
const WHERE_COMPLEXITY_FACTOR = 2
const PAGE_LIMIT = 100
```

File: packages/keystone/rateLimiting/plugin.js [hunk 2]
```diff
@@ -0,0 +1,196 @@
+const ms = require('ms')
```
Comment (sitozzz) [inline packages/keystone/rateLimiting/plugin.js:1]: Does you really need package that converts to milliseconds? There are only two usage of it...

File: packages/keystone/rateLimiting/plugin.js [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+                const { isAuthed, key } = extractQuotaKeyFromRequest(requestContext)
+                const maxQuota = isAuthed ? this.#authedQuota : this.#nonAuthedQuota
+
+                /** @type {string | null} */
+                const currentValue = await this.#redisClient.get(key)
+                const usedQuota = parseInt(currentValue) || 0
+
+                if (usedQuota + requestComplexity > maxQuota) {
+                    // TODO: Throw error instead here
```
Comment (sitozzz) [inline packages/keystone/rateLimiting/plugin.js:180]: Maybe add some log or metric hit instead of error for now?

**Review by sitozzz — COMMENTED** [review]
I can't figure out how will you get weights adjustments without collecting special data or logs. You may use logger or `metrics` package

File: packages/keystone/rateLimiting/constants.js (hunk 1 above)
Comment (SavelevMatthew) [reply packages/keystone/rateLimiting/constants.js:8]: Yep, you can override all of it, by specifying `new ApolloRateLimitPlugin(keystone, { pageLimit: 10 })` or so

File: packages/keystone/rateLimiting/plugin.js (hunk 2 above)
Comment (SavelevMatthew) [reply packages/keystone/rateLimiting/plugin.js:1]: In a perfect world (out of MVP) I probably want to override values on `per-query/mutation` level...
So if I do something like...
```
fields: ...
hooks: ...
rateLimit: {
    create: {
        window: "5min"
        quota: 25
    }
}
```

Plugin will generate separate reeds key for this mutation, that's the reason I've introduced this npm package. Quality of life

File: packages/keystone/rateLimiting/plugin.js (hunk 3 above)
Comment (SavelevMatthew) [reply packages/keystone/rateLimiting/plugin.js:180]: It's already added. I've added all calculated info to `context.req`. And pino plugin will extract it later for Kibana to analyse

**Review by sitozzz — APPROVED** [review]

File: packages/keystone/rateLimiting/plugin.js [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+                    return
```
Comment (pahaz) [inline packages/keystone/rateLimiting/plugin.js:181]: ?

File: packages/keystone/rateLimiting/plugin.js [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+                }
+
+                if (currentValue === null) {
```
Comment (pahaz) [inline packages/keystone/rateLimiting/plugin.js:184]: It's not atomic

**Pull request merged**